nsxt/resource_nsxt_upgrade_prepare.go

```diff
@@ -0,0 +1,378 @@
+/* Copyright © 2023 VMware, Inc. All Rights Reserved.
+   SPDX-License-Identifier: MPL-2.0 */
+
+package nsxt
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx"
+	nsxModel "github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/model"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/upgrade"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/upgrade/bundles"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/upgrade/eula"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/upgrade/pre_upgrade_checks"
+)
+
+var precheckComponentTypes = []string{"EDGE", "HOST", "MP"}
+
+func resourceNsxtUpgradePrepare() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceNsxtUpgradePrepareCreate,
+		Read:   resourceNsxtUpgradePrepareRead,
+		Update: resourceNsxtUpgradePrepareUpdate,
+		Delete: resourceNsxtUpgradePrepareDelete,
+
+		Schema: map[string]*schema.Schema{
+			"upgrade_bundle_url": {
+				Type:        schema.TypeString,
+				Description: "URL of the NSXT Upgrade bundle",
+				Required:    true,
+			},
+			"precheck_bundle_url": {
+				Type:        schema.TypeString,
+				Description: "URL of the NSXT Upgrade precheck bundle (Only applied to NSXT version >= 4.1.1)",
+				Optional:    true,
+			},
+			"version": {
+				Type:        schema.TypeString,
+				Description: "Version for the upgrade",
+				Required:    true,
+			},
+			"accept_user_agreement": {
+				Type:        schema.TypeBool,
+				Description: "Whether to accept the user agreement",
+				Required:    true,
+			},
+			"failed_prechecks": {
+				Type:        schema.TypeList,
+				Description: "List of failed prechecks for the upgrade, only include warnings, if precheck failed with error then the resource creation will fail",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:        schema.TypeString,
+							Description: "ID of failed precheck",
+							Computed:    true,
+						},
+						"message": {
+							Type:        schema.TypeString,
+							Description: "Message of the failed precheck",
+							Computed:    true,
+						},
+					},
+				},
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceNsxtUpgradePrepareCreate(d *schema.ResourceData, m interface{}) error {
+	id := d.Id()
+	if id == "" {
+		id = newUUID()
+	}
+	err := prepareForUpgrade(d, m)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", id, err)
+	}
+	d.SetId(id)
+	return resourceNsxtUpgradePrepareRead(d, m)
+}
+
+func prepareForUpgrade(d *schema.ResourceData, m interface{}) error {
+	// 1. Upload upgrade bundle and wait for upload to complete
+	err := uploadPrecheckAndUpgradeBundle(d, m)
```

Comment:
I assumed we would skip this step if bundles are already uploaded (and haven't changed since), but I don't see this logic implemented. Is there a reason to upload again if bundle status is already ok?

Reply:
Added a step to check upgrade summary, it will compare the existing bundle file name to the bundle file name from url, if they are equal then skip the upload process.